funsor/gaussian.py

```diff
@@ -114,15 +114,31 @@ def as_tensor(self):
         prototype = next(iter(self.parts.values()))
         for i in _find_intervals(self.parts.keys(), self.shape[-1]):
             if i not in self.parts:
-                self.parts[i] = ops.new_zeros(prototype, self.shape[:-1] + (i[1] - i[0],))
+                self.parts[i] = ops.new_zeros(prototype, (i[1] - i[0],))
 
         # Concatenate parts.
-        parts = [v for k, v in sorted(self.parts.items())]
+        parts = [ops.expand(v, self.shape[:-1] + v.shape[-1:]) for k, v in sorted(self.parts.items())]
         result = ops.cat(-1, *parts)
         if not get_tracing_state():
             assert result.shape == self.shape
         return result
 
+    def __add__(self, other):
+        assert isinstance(other, BlockVector)
```

Comment:
I think we'll also want either (1) an assertion that the two block forms are compatible, or (2) a brach to convert both `.as_tensor()` and add the results. Two block forms are incompatible if any of their slices overlap, e.g. `[0:3],[3:10]` versus `[0:4],[4:10]`.

Reply:
I remember that I was worried about this issue too but later found that there is no test for it. So I assumed that each block corresponds to a variable and we only add two BlockVectors when their common variables have consistent dimensions. If that is true, I'll add an assertation. Otherwise, you are right that we need a branch.

Reply:
Either way, we need a check for that consistency. So adding a branch has no harm after all. I'll do it.